# I94 Immigration Data Prob

### First let's start spark and load the data:

In [3]:
from pyspark.sql import SparkSession

spark = SparkSession.builder.enableHiveSupport().getOrCreate();

df = spark.read.parquet('output/immigrations/')
df.createOrReplaceTempView("immigration")
df.show()

+------------------+----------------------+---------+------------+------------+----------+--------------+---+-------------+-----+----------+-----------------+----------+------------+--------------+-----------+---------------------------+----------+------------+----------+-------+----------------+-------------+---------+----+-----+
|birth_country_code|residence_country_code|port_code|arrival_date|arrival_mode|state_code|departure_date|age|visa_category|count|date_added|visa_issuing_post|occupation|arrival_flag|departure_flag|update_flag|match_arrival_departure_fag|birth_year|allowed_date|ins_number|airline|admission_number|flight_number|visa_type|year|month|
+------------------+----------------------+---------+------------+------------+----------+--------------+---+-------------+-----+----------+-----------------+----------+------------+--------------+-----------+---------------------------+----------+------------+----------+-------+----------------+-------------+---------+----+-----+
|

AttributeError: 'DataFrame' object has no attribute 'countries'

### Now let's start from simple questions:

#### Q1. Which country does most immigrations come from?

In [7]:
df_countries = spark.read.csv('output/country_codes.csv',header=True)
df_countries.createOrReplaceTempView("country_codes")
df_countries.show()

dfr = spark.sql(
    '''
    select cc.country,count(1)
    from immigration i
    join country_codes cc on cc.code = i.residence_country_code
    group by cc.country
    order by count(1) desc
    '''
)
dfr.show()

+----+--------------------+
|code|             country|
+----+--------------------+
| 582|MEXICO Air Sea, a...|
| 236|         AFGHANISTAN|
| 101|             ALBANIA|
| 316|             ALGERIA|
| 102|             ANDORRA|
| 324|              ANGOLA|
| 529|            ANGUILLA|
| 518|     ANTIGUA-BARBUDA|
| 687|           ARGENTINA|
| 151|             ARMENIA|
| 532|               ARUBA|
| 438|           AUSTRALIA|
| 103|             AUSTRIA|
| 152|          AZERBAIJAN|
| 512|             BAHAMAS|
| 298|             BAHRAIN|
| 274|          BANGLADESH|
| 513|            BARBADOS|
| 104|             BELGIUM|
| 581|              BELIZE|
+----+--------------------+
only showing top 20 rows

+--------------------+--------+
|             country|count(1)|
+--------------------+--------+
|      UNITED KINGDOM|  368421|
|               JAPAN|  249167|
|          CHINA, PRC|  185609|
|              FRANCE|  185339|
|MEXICO Air Sea, a...|  179603|
|             GERMANY|  156613|
|         SOUT

conclusion: top incoming countries are:
|UNITED KINGDOM      |  368421|
|               JAPAN|  249167|
|          CHINA, PRC|  185609|
|              FRANCE|  185339|
|MEXICO Air Sea, a...|  179603|
|             GERMANY|  156613|
|         SOUTH KOREA|  136312|
|              BRAZIL|  134907|
|           AUSTRALIA|  112407|
|               INDIA|  107193

